credentials/src/main/kotlin/web5/sdk/credentials/model/VcDataModel.kt

```diff
@@ -0,0 +1,473 @@
+   *
+   * @return A Map containing key-value pairs representing the properties of the [VcDataModel] instance.
+   */
+  public fun toMap(): Map<String, Any> = mutableMapOf<String, Any>().apply {
+    id?.also { put("id", it.toString()) }
+    put("@context", context.map { uri -> uri.toString() })
+    put("type", type)
+    put("issuer", issuer.toString())
+    put("issuanceDate", DATE_FORMAT.format(issuanceDate))
+    expirationDate?.also { put("expirationDate", DATE_FORMAT.format(it)) }
+    put("credentialSubject", credentialSubject.toMap())
+    credentialSchema?.also { put("credentialSchema", it.toMap()) }
+    credentialStatus?.also { put("credentialStatus", it.toMap()) }
+  }
+
+  public companion object {
+    /**
+     * Builds a new [CredentialSubject] instance from a map of properties.
+     *
+     * @param map A map representing the [CredentialSubject]'s properties.
+     * @return An instance of [CredentialSubject] constructed from the provided map.
+     */
+    public fun fromMap(map: Map<String, Any>): VcDataModel {
+      require(map.containsKey("issuer")) { "Issuer is required" }
+      require(map.containsKey("issuanceDate")) { "IssuanceDate is required" }
+      require(map.containsKey("credentialSubject")) { "CredentialSubject is required" }
+
+      return VcDataModel(
+        id = (map["id"] as? String)?.let { URI.create(it) },
+        context = (map["@context"] as? List<*>)?.mapNotNull {
+          (it as? String)?.let { str -> URI.create(str) }
+        } ?: listOf(),
+        type = map["type"] as? List<String> ?: listOf(),
+        issuer = URI.create(map["issuer"] as String),
+        issuanceDate = DATE_FORMAT.parse(map["issuanceDate"] as String),
+        expirationDate = (map["expirationDate"] as? String)?.let { DATE_FORMAT.parse(it) },
+        credentialSubject = CredentialSubject.fromMap(map["credentialSubject"] as Map<String, Any>),
+        credentialSchema = (map["credentialSchema"] as? Map<String, Any>)?.let { CredentialSchema.fromMap(it) },
+        credentialStatus = (map["credentialStatus"] as? Map<String, Any>)?.let { BitstringStatusListEntry.fromMap(it) }
+      )
+    }
+
+    /**
+     * Parses a JSON string to create an instance of [VcDataModel].
+     *
+     * @param jsonString The JSON string representation of a [VcDataModel].
+     * @return An instance of [VcDataModel].
+     */
+    public fun fromJsonObject(jsonString: String): VcDataModel {
+      val mapper = jacksonObjectMapper()
+      return mapper.readValue(jsonString, object : TypeReference<VcDataModel>() {})
+    }
+  }
+}
+
+/**
+ * The [CredentialSubject] represents the value of the credentialSubject property as a set of objects containing
+ * properties related to the subject of the verifiable credential.
+ */
+public data class CredentialSubject(
+  val id: URI? = null,
+  val additionalClaims: Map<String, Any> = emptyMap()
+) {
+
+  /**
+   * Builder for [CredentialSubject].
+   */
+  public class Builder {
```

Comment:
nit: u can nest Builder class inside the credentialsubject class like this. i like this a bit better, encapsulation and all that

https://github.com/TBD54566975/web5-kt/pull/248/files#diff-81d9b32e2a369effdea19a4888a66a9d5a44b0b70dcf27cce32067dcb86efed9

Reply:
perfect, I'll do this